rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+performance doesn't matter:
+
+```bash
+vast from s3://aws |
+  vast read json |
+  vast group-by |
+  vast summarize |
+  vast write feather |
+  vast to /path/to/file.feather
+```
+
+But let's take a step back and just assume that we add a new `exec` command that
+just executes a pipeline.
+
+Then we can model the ingestion as follows: load data via stdin, push into a
+parser, then send off to a remote VAST node. For example, we would rewrite
+`import zeek` as:
+
+```bash
+vast exec 'from - | parse zeek | to vast://1.2.3.4'
+```
+
+Likewise, `export json EXPRESSION` would become:
+
+```bash
+vast exec 'from vast://1.2.3.4 |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The beauty is that we can almost write the same without requiring a VAST node
+and execute *in situ*:
+
+```bash
+vast exec 'from - |
+           where EXPRESSION |
+           write json |
+           to -'
+```
+
+The only difference being `-` here.
+
+The VAST node basically becomes a **pipeline manager** with persistent state. We
+can keep everything as is, e.g., start a node:
+
+```bash
+vast start -e 1.2.3.4
+```
+
+There are of course a bunch of convenience things we can do, like assuming `from
+-` and `to -` being the default. This would make the above look like this:
+
+```bash
+vast exec 'where EXPRESSION |
+           write json
+```
+
+Side effect: we would get our envisioned convert operator almost for free:
+
+```bash
+vast exec 'parse zeek | print csv'
+# Obviates the need for:
+vast exec 'convert zeek csv'
+```
+
+Another UX improvement would be assuming that we interact with a VAST node. Here
+we could have dedicate commands the prepend or append a pipeline operator:
+
+- `vast pull` → `from vast://...`
+- `vast push` → `to vast://...`
+
+An `export` would then become:
+
+```bash
+vast pull 'where EXPRESSION | print json'
+# Same with exec:
+vast exec 'from vast://.... | where EXPRESSION | print json'
+# Same with export:
+vast export json EXPRESSION
+```
+
+An `import` degenerates to
+
+```bash
+vast push 'parse zeek | where EXPRESSION'
+# Same with exec
+vast exec 'parse zeek | where EXPRESSION | to vast://...'
+# Same with import
+vast import zeek EXPRESSION
+```
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the middle.
+It's clean and simple.
+
+But this may be too restrictive. Let's assume for a moment that input and output
+are dynamically typed. Then an operator must define its input and output types,
+e.g., `from<In, Out>`. This can come in handy for working with matchers:
+
+Here, we build a matcher, with read having the type `read<Bytes, Arrow>`:
+
+```bash
+vast exec 'read csv | matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+The type of `matcher build` would be `<Arrow, Flatbuffer>`. Also assume `matcher
+load` yields `<Flatbuffer, Void>` where `Void` is "no output, just a side
+effect".
```

Comment:
From a technical perspective, we need some type-erased variant that covers _all_ possible alternatives for input and output types. This should probably be covered in a separate technical section.

Reply:
I think that's fair to assume that there will be a sum type for input and output, potentially even the same. But that's the C++ layer. I'm not sure if I want to go here in the proposal.

Reply:
I think the proposal would be incomplete without it. This necessarily has to change the interface of pipeline operators, and that's a change that needs to be looked at carefully.

Reply:
We can certainly have a draft API, but that follows after we've agreed on the top-level mechanics and behavior. I would prefer to not go there yet, but rather wait for further feedback.

Reply:
There exists a table now, closing this out. Feel free to comment in the table if that needs rework.